cadence/contracts/bridge/FlowEVMBridge.cdc

```diff
@@ -168,6 +188,144 @@ contract FlowEVMBridge : IFlowEVMNFTBridge, IFlowEVMTokenBridge {
         self.deployDefiningContract(evmContractAddress: address)
     }
 
+    /// Registers a custom cross-VM NFT implementation, allowing projects to integrate their Cadence & EVM contracts
+    /// such that the VM bridge facilitates movement between VMs as the integrated implementations.
+    ///
+    /// @param type: The NFT Type to register as cross-VM NFT
+    /// @param fulfillmentMinter: The optional NFTFulfillmentMinter Capability. This parameter is required for
+    ///     EVM-native NFTs
+    /// @param feeProvider: A reference to a FungibleToken Provider from which the bridging fee is withdrawn in $FLOW
+    ///
+    access(all)
+    fun registerCrossVMNFT(
+        type: Type,
+        fulfillmentMinter: Capability<auth(FlowEVMBridgeCustomAssociationTypes.FulfillFromEVM) &{FlowEVMBridgeCustomAssociationTypes.NFTFulfillmentMinter}>?,
+        feeProvider: auth(FungibleToken.Withdraw) &{FungibleToken.Provider}
+    ) {
+        pre {
+            FlowEVMBridgeUtils.typeAllowsBridging(type):
+            "This Cadence Type \(type.identifier) is currently opted-out of bridge onboarding"
+            type.isSubtype(of: Type<@{NonFungibleToken.NFT}>()):
+            "The provided Type \(type.identifier) is not an NFT - only NFTs can register as cross-VM"
+            !type.isSubtype(of: Type<@{FungibleToken.Vault}>()):
+            "The provided Type \(type.identifier) is also a FungibleToken Vault - only NFTs can register as cross-VM"
+            !FlowEVMBridgeConfig.isCadenceTypeBlocked(type):
+            "Type \(type.identifier) has been blocked from onboarding"
+            FlowEVMBridgeUtils.isCadenceNative(type: type):
+            "Attempting to register a bridge-deployed NFT - cannot update a bridge-defined asset"
```

Comment:
I think I asked this somewhere else also, but just to clarify, there is no way to update an EVM-native asset from before this upgrade to a cross VM NFT? I can't think of a way to get around Solidity's upgrade restrictions myself, but just wondering if there is some way it could be possible. Probably out of the scope of this

Reply:
```suggestion
            FlowEVMBridgeUtils.isCadenceNative(type: type):
            "Attempting to register an EVM-native bridge-deployed NFT - cannot update a bridge-defined asset that originated in EVM because EVM smart contracts are not upgradable"
```
Either way, the error message confused me at first. It wasn't immediately clear that being Cadence Native meant it isn't updatable. Maybe this could be the error message.

Reply:
Good point, this message is kind of vague. The issue here isn't so much that it's EVM-native as much as we won't allow projects to alter bridge-defined associations. Technically, if an EVM project deployed using upgradeable patterns and onboarded the proxy address, the developer could update their contract implementation in conformance with cross-VM standards and register their own Cadence contract. But in such a case they would register via their Cadence type, not the bridged Type that was previously associated with their EVM contract.
By including this pre-condition, I'm trying to catch a failure later in scope and explicitly call out that updating bridge-defined types is disallowed. I'll try to clarify the error message.

Reply:
Updated in [696a37e](https://github.com/onflow/flow-evm-bridge/pull/173/commits/696a37e57d640a1a6a692436d64ef609f34e62f4)